automod/countstore/countstore_mem.go

```diff
@@ -0,0 +1,64 @@
+package countstore
+
+import (
+	"context"
+
+	"github.com/puzpuzpuz/xsync/v3"
+)
+
+type MemCountStore struct {
+	// Counts is keyed by a string that is a munge of "{name}/{val}[/{period}]",
+	// where period is either absent (meaning all-time total)
+	// or a string describing that timeperiod (either "YYYY-MM-DD" or that plus a literal "T" and "HH").
+	//
+	// (Using a values for `name` and `val` with slashes in them is perhaps inadvisable, as it may be ambiguous.)
+	Counts         *xsync.MapOf[string, int]
+	DistinctCounts *xsync.MapOf[string, *xsync.MapOf[string, bool]]
+}
+
+func NewMemCountStore() MemCountStore {
+	return MemCountStore{
+		Counts:         xsync.NewMapOf[string, int](),
+		DistinctCounts: xsync.NewMapOf[string, *xsync.MapOf[string, bool]](),
+	}
+}
+
+func (s MemCountStore) GetCount(ctx context.Context, name, val, period string) (int, error) {
+	v, ok := s.Counts.Load(periodBucket(name, val, period))
+	if !ok {
+		return 0, nil
+	}
+	return v, nil
+}
+
+func (s MemCountStore) Increment(ctx context.Context, name, val string) error {
+	for _, p := range []string{PeriodTotal, PeriodDay, PeriodHour} {
+		k := periodBucket(name, val, p)
+		s.Counts.Compute(k, func(oldVal int, _ bool) (int, bool)  {
+			return oldVal+1, false
+		})
+	}
+	return nil
+}
+
+func (s MemCountStore) GetCountDistinct(ctx context.Context, name, bucket, period string) (int, error) {
+	v, ok := s.DistinctCounts.Load(periodBucket(name, bucket, period))
+	if !ok {
+		return 0, nil
+	}
+	return v.Size(), nil
+}
+
+func (s MemCountStore) IncrementDistinct(ctx context.Context, name, bucket, val string) error {
+	for _, p := range []string{PeriodTotal, PeriodDay, PeriodHour} {
+		k := periodBucket(name, bucket, p)
+		s.DistinctCounts.Compute(k,func(nested *xsync.MapOf[string, bool], _ bool) (*xsync.MapOf[string, bool], bool)  {
```

Comment:
does the nested map need to be a sync primitive as well? my intuition is that the sync control of the outer map should protect the inner map.

this isn't a review-blocking issue, just curious

Reply:
...Okay, that's a good question.  I've been nerdsnipped and I'm actually still a bit unsure.

My theoretical understanding: I don't think it would be safe.  (And not just like "aw that's a minor inaccuracy", like, I think the race detector _should_ flag it.)

- I read the source of this `syncx` package:
  - This map _does_ use some mutexing during writes.
  - And that mutexing _does_ cover the compute function callback body.
  - It does _not_ use mutexting during reads.  (Reads are done purely with atomic load operations.)
  - (And obviously, once you've done a read, you've... got the value; even if there was mutexing during the read (and there's not), it couldn't still be in effect after you hold the value.)
- So do we do anything on the value after a read?  Yes, we do have a read where we extract the inner map... and then call `.Size()` on it (or would call `len()` if it was a native map).
- Is that operation safe?
  - `xsync`'s `.Size()`: yes, defintely.  It's another atomic read.
  - native `len()`:  No. `len()` of native maps is considered a read for race detector purposes.  Performing one outside of a mutex (even if all writes are mutexed) _will_ trigger the race detector.  (I even wrote a little side test to be sure -- separate from our nested maps situation here.  Yep, easy to trigger.)

Practically verifying the whole thing: I wrote a quick variant with a plain native map on the inner, and... I'm unable to get the race detector to flag it, nor get the test to fail.

:exploding_head: 

Overall: I'm suspicious my testing is somehow not aggressive enough, rather than confident that a plain map on the inner is fully safe.